pkg/sql/tests/sysbench_test.go

```diff
@@ -0,0 +1,624 @@
+	sysbenchExecuteRange(s.DistinctRange, sysbenchDistinctRanges, rng)
+}
+
+func sysbenchExecuteIndexUpdates(s sysbenchDriver, rng *rand.Rand) {
+	t := randTableNum(rng)
+	for range sysbenchIndexUpdates {
+		s.IndexUpdate(t, randRowID(rng))
+	}
+}
+
+func sysbenchExecuteNonIndexUpdates(s sysbenchDriver, rng *rand.Rand) {
+	t := randTableNum(rng)
+	for range sysbenchNonIndexUpdates {
+		s.NonIndexUpdate(t, randRowID(rng), randCValue(rng))
+	}
+}
+
+func sysbenchExecuteDeleteInserts(s sysbenchDriver, rng *rand.Rand) {
+	t := randTableNum(rng)
+	for range sysbenchDeleteInserts {
+		id := randRowID(rng)
+		k := randKValue(rng)
+		c := randCValue(rng)
+		pad := randPadValue(rng)
+		s.DeleteInsert(t, id, k, c, pad)
+	}
+}
+
+// sysbenchWorkload executes a single transaction of a sysbench workload.
+type sysbenchWorkload func(sysbenchDriver, *rand.Rand)
+
+func sysbenchOltpReadOnly(s sysbenchDriver, rng *rand.Rand) {
+	s.Begin()
+	sysbenchExecutePointSelects(s, rng)
+	sysbenchExecuteSimpleRanges(s, rng)
+	sysbenchExecuteSumRanges(s, rng)
+	sysbenchExecuteOrderRanges(s, rng)
+	sysbenchExecuteDistinctRanges(s, rng)
+	s.Commit()
+}
+
+func sysbenchOltpWriteOnly(s sysbenchDriver, rng *rand.Rand) {
+	s.Begin()
+	sysbenchExecuteIndexUpdates(s, rng)
+	sysbenchExecuteNonIndexUpdates(s, rng)
+	sysbenchExecuteDeleteInserts(s, rng)
+	s.Commit()
+}
+
+func sysbenchOltpReadWrite(s sysbenchDriver, rng *rand.Rand) {
+	s.Begin()
+	sysbenchExecutePointSelects(s, rng)
+	sysbenchExecuteSimpleRanges(s, rng)
+	sysbenchExecuteSumRanges(s, rng)
+	sysbenchExecuteOrderRanges(s, rng)
+	sysbenchExecuteDistinctRanges(s, rng)
+	sysbenchExecuteIndexUpdates(s, rng)
+	sysbenchExecuteNonIndexUpdates(s, rng)
+	sysbenchExecuteDeleteInserts(s, rng)
+	s.Commit()
+}
+
+func BenchmarkSysbench(b *testing.B) {
+	defer log.Scope(b).Close(b)
+	for _, sysFn := range []func(context.Context, *testing.B) (sysbenchDriver, func()){
+		newSysbenchSQL,
+		newSysbenchKV,
+		// TODO(nvanbenschoten): add a pebble-level implementation.
+	} {
+		sysTyp := runtime.FuncForPC(reflect.ValueOf(sysFn).Pointer()).Name()
+		sysTyp = strings.TrimPrefix(sysTyp, "github.com/cockroachdb/cockroach/pkg/sql/tests_test.newSysbench")
+		b.Run(sysTyp, func(b *testing.B) {
+			for _, opFn := range []sysbenchWorkload{
+				sysbenchOltpReadOnly,
+				sysbenchOltpWriteOnly,
+				sysbenchOltpReadWrite,
+			} {
+				opTyp := runtime.FuncForPC(reflect.ValueOf(opFn).Pointer()).Name()
+				opTyp = strings.TrimPrefix(opTyp, "github.com/cockroachdb/cockroach/pkg/sql/tests_test.sysbench")
+				b.Run(opTyp, func(b *testing.B) {
+					defer func() {
+						if r := recover(); r != nil {
+							b.Fatalf("%+v", r)
+						}
+					}()
+
+					ctx := context.Background()
+					sys, cleanup := sysFn(ctx, b)
+					defer cleanup()
+
+					rng := rand.New(rand.NewSource(0))
+					sys.prep(rng)
+
+					b.ResetTimer()
+					for i := 0; i < b.N; i++ {
+						opFn(sys, rng)
+					}
+				})
+			}
+		})
+	}
+}
+
+func try0(err error) {
+	if err != nil {
```

Comment:
how confident are we that we don't have to handle errors here? I know, single node cockroach usually doesn't have any, but can we at least fail gracefully? For example, stash the error in an atomic and check that at conclusion time. After all, this stuff will run in CI (and our ambition may be to put it squarely into the critical path once we can justify it).